src/server/pkg/storage/fileset/dir_inserter.go

```diff
@@ -0,0 +1,79 @@
+package fileset
+
+import (
+	"context"
+	"io"
+	"path"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/tar"
+)
+
+type dirInserter struct {
+	x FileSource
+}
+
+func NewDirInserter(x FileSource) FileSource {
+	return &dirInserter{x: x}
+}
+
+func (s *dirInserter) Iterate(ctx context.Context, cb func(File) error, stopBefore ...string) error {
```

Comment:
This seems fine for the time being, but I think we may want this functionality to be on the write side. Maybe a wrapper for the fileset writer. This would make it so that both the copy file process and the serialization of the in memory filesets can have one implementation that handles creating directories implicitly.

Reply:
We limit the usefulness of a transformation if it is on the write side.  Writing goes to somewhere we can't get at.  Transformations on the read side we can reuse.  The best example right now is GetTar. If we want to return directory entries in that, we would need to wrap this around the `FileSource` that is currently returned.

Reply:
Thinking more about this.  If the "UnorderedWriter" was able to wrap its internal map with something that looked like a `FileSource` we could read from it and apply the directory inserter before writing it to the "OrderedWriter".

Reply:
I think this is more or less what we want
```go
package fileset

import "sort"

type Buffer map[string]File

func (b Buffer) Iterate(ctx context.Context, cb func(f File) error) error {
        keys := []string{}
        for k := range b {
                keys = append(keys, key)
        }
        sort.Strings(keys)
        for k := range keys {
                f = b[k]
                if err := cb(f); err != nil {
                        return err
                }
        }
        return nil
}
```

Reply:
This makes sense to me. The main thing is that I want that functionality to be captured in one place, and having it be a read abstraction could potentially be useful if we come up with another use case for inserting directories that is not doing so just to satisfy the invariant we have on ordered writes.